File: 6week/template/4.maximum-split-of-positive-even-integers.py
## https://leetcode.com/problems/maximum-split-of-positive-even-integers/

from typing import List

class Solution:
    def maximumEvenSplit(self, finalSum: int) -> List[int]:
        if finalSum % 2 == 1:
            return []
        i, sum = 0, 0
        answer = []
        while sum + i < finalSum:
            i += 2
            sum += i
            answer.append(i)
        if sum == finalSum:
            return answer
        last = answer.pop()
        answer.append(finalSum-sum+last)
        return answer

print(Solution().maximumEvenSplit(28))

File: 6week/남규/4. Maximum Split of Positive Even Integers.cpp
class Solution {
public:
    vector<long long> maximumEvenSplit(long long finalSum) {
        vector<long long> ans;
        
        if(finalSum % 2 == 1) return ans;
        
        for(long long even = 2; finalSum; even += 2){
            if(even == finalSum){
                ans.push_back(even);
                break;
            }
            
            if(finalSum - even >= even + 2){
                finalSum -= even;
                ans.push_back(even);
            }
        }
        return ans;
    }
};
